### PR TITLE
[web] flutter.js initialization with ui.webOnlyWarmupEngine

### LINKED ISSUE #99773: [web] Let developers control the bootstrap of a Flutter Web application
Currently, Flutter Web apps launch as soon as their `main.dart.js` is loaded and parsed. This makes it harder for web developers to integrate Flutter web apps into their sites (for example, it's hard to create a progress bar while the app is loading, or prevent a flutter app from using resources until it's really needed).

Let's introduce a Flutter Web bootstrap API that lets developers control how (when) the bulk of the Flutter web app is downloaded, and run.

## Use case

* Users want to control *when* a Flutter web app starts utilizing computing resources (network, processor) on their web app.
* Ability to create a somewhat accurate "loading" bar.
* Increase Lighthouse scores by converting Flutter web apps into something that can be asynchronously loaded and run.

## Proposal

* Design doc: [📄 Programmatic Bootstrap of a Flutter Web App](https://docs.flutter.dev/go/web-programmatic-bootstrap)
* PoC: [Demo](https://dit-tests.web.app)

### PR DESCRIPTION
**Pull request by ditman**
This PR changes the initialization of a flutter web app to use the `ui.webOnlyWarmupEngine` API introduced here:

* https://github.com/flutter/engine/pull/32017

in order to yield full control to the programmer over the initialization of a Flutter Web app.

* **Fixes: https://github.com/flutter/flutter/issues/99773**
* Supersedes: https://github.com/flutter/flutter/pull/95379

Currently, Flutter Web apps start as soon as possible, and take over the whole screen. With this change, the startup of an App can be split in three stages (four, if we count the cleanup stage too):

* Injecting the main.dart.js and (optionally) the serviceworker
* Initializing the engine
* Running the flutter app
* Cleaning the app (currently not implemented)

To achieve this, some JS-interop and Promise-wrangling is needed. The new workflow is as follows:

* The user injects flutter.js in their index.html, and once that script is loaded, calls `_flutter.loader.loadMainDartJs`
* flutter.js injects main.dart.js, and sets up a `_flutter.loader.didLoadMainDartJs` callback. (This callback must be in a hardcoded location, so Flutter loaded from a script knows where to call)
  * If the callback is not present (old index.html, not using flutter.js):
    * The engine immediately initializes the app, as it used to do earlier. This ensures backwards compatibility.
  * If the callback is present:
    * The engine calls it with an `EngineInitializer` object that can be used by the programmer.
      * This resolves the promise returned by `loadMainDartJs` in step 1.
    * From this moment on, the loading of the application is a sequence of Promise-based objects that let the user walk through the three (four) stages described above:

    ```
    loadMainDartJs -> EngineInitializer.initializeEngine -> AppRunner.runApp -> AppCleaner.cleanApp (optional)
    ```

### Demo

The new API above allows users to write their own (informative) splash screens, for example:

* **https://dit-bootstrap-tests.web.app** (red circles should turn green as the app loads. The button only becomes enabled once the app is ready to start. The app only starts when the button is pressed).

### Backwards compatibility

This is currently backwards compatible with old versions of index.html. If the flutter init script calls `ui.webOnlyWarmUpEngine`, it checks if the flutter loader callback is setup or not.

### Testing

* Unit tests added to assert the existence and contents of the new flutter.js file.
* Minor updates to existing unit tests that were asserting the contents of the old wrapper script (auto-generated)
* general.shard of tests passes.

Waiting to update this branch with the latest from the engine, to detect potential failures in integration tests between the engine and the framework, even though the engine change should be fully backwards-compatible.

### Manual testing

Clone [this PR](https://github.com/flutter/engine/pull/32017) into your flutter/engine fork, and build it.

Clone this PR into your flutter/flutter fork, and ensure that your `flutter` binary is the one coming from your Github checkout.

With the new version of flutter on your box, you'll be able to create a new app normally by using `flutter create`. That should come with the new flutter.js and index.html templates.

In order for the flutter_tool to rebuild, you'll need to delete its cache; you can do it with:

```
rm [YOUR_FLUTTER_CLONE]/bin/cache/flutter_tools.*
```

Once the engine and the tools are rebuilt, you can just do something similar to:

```
flutter clean && flutter run --local-engine=host_debug_unopt -d chrome --web-port 7357
```

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/templates/app_shared/web/index.html.tmpl [hunk 1]
```diff
@@ -31,74 +31,34 @@
...
+    window.addEventListener('load', function(ev) {
+      // Download main.dart.js
+      _flutter.loader.loadMainDartJs({
+        entrypointUrl: 'main.dart.js',
+        serviceWorker: {
+          serviceWorkerVersion: serviceWorkerVersion,
+          timeoutMillis: 4000
+        }
+      }).then(function(engineInitializer) {
+        return engineInitializer.initializeEngine({
```
Comment (yjbanov) [inline packages/flutter_tools/templates/app_shared/web/index.html.tmpl:53]: Let's remove the config argument until we have something more concrete to say about it.

File: packages/flutter_tools/templates/app_shared/web/index.html.tmpl [hunk 2]
```diff
@@ -31,74 +31,34 @@
...
+          // Potential engine config
+        });
+      }).then(function(appRunner) {
+        return appRunner.runApp({
+          // App startup parameters
```
Comment (yjbanov) [inline packages/flutter_tools/templates/app_shared/web/index.html.tmpl:58]: ditto: let's remove config arg for now.

File: packages/flutter_tools/templates/app_shared/web/flutter.js.tmpl [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    // Resolver for the pending promise returned by loadMainDartJs.
+    _didCreateEngineInitializerResolve = null;
+
+    /**
+     * Initializes the main.dart.js with/without serviceWorker.
+     * @param {*} options
+     * @returns a Promise that will eventually resolve with an EngineInitializer,
+     * or will be rejected with the error caused by the loader.
+     */
+    loadMainDartJs(options) {
```
[the remark this answers is not in the document]
Comment (ditman) [reply packages/flutter_tools/templates/app_shared/web/flutter.js.tmpl:34]: Renamed

File: packages/flutter_tools/templates/app_shared/web/index.html.tmpl [hunk 4]
```diff
@@ -31,74 +31,34 @@
...
-          if (!scriptLoaded) {
-            console.warn(
-              'Failed to load app from service worker. Falling back to plain <script> tag.',
-            );
-            loadMainDartJs();
-          }
-        }, 4000);
+    window.addEventListener('load', function(ev) {
+      // Download main.dart.js
+      _flutter.loader.loadMainDartJs({
```
[the remark this answers is not in the document]
Comment (ditman) [reply packages/flutter_tools/templates/app_shared/web/index.html.tmpl:46]: It'd create a fancy explosion, if I didn't rename this one too :P

File: packages/flutter_tools/templates/app_shared/web/index.html.tmpl (hunk 1 above)
Comment (ditman) [reply packages/flutter_tools/templates/app_shared/web/index.html.tmpl:53]: Removed pretty much all the arguments that are defaulted by flutter.js

**Review by yjbanov — APPROVED** [review]
lgtm

**Comment by ditman** [comment]
Until the engine PR lands and rolls to flutter (and this branch is updated), these failures will happen:

```
Building flutter tool...
org-dartlang-app:/web_entrypoint.dart:12:12: Error: Method not found: 'webOnlyWarmupEngine'.
  await ui.webOnlyWarmupEngine(
           ^^^^^^^^^^^^^^^^^^^
Failed to compile application.
```

Potential other failures may happen even after the update, even though the changes are fairly backwards-compatible.

File: packages/flutter_tools/lib/src/build_system/targets/web.dart [hunk 5]
```diff
@@ -120,48 +121,54 @@ class WebEntrypointTarget extends Target {
...
-    } else {
-      contents = '''
-// @dart=${languageVersion.major}.${languageVersion.minor}
-
-import 'dart:ui' as ui;
+    }
 
-import '$mainImport' as entrypoint;
+    final String contents = <String>[
+        '// @dart=${languageVersion.major}.${languageVersion.minor}',
```
Comment (ditman) [inline packages/flutter_tools/lib/src/build_system/targets/web.dart:139]: TODO: Export this to a different `lib/src/web/web_entrypoint.dart` file, so code is shared between the `flutter build` and `run` commands.

**Comment by ditman** [comment]
Removed manual roll, and rebased with proper master.

**Comment by fangkyi03** [comment]
delete Platform.isAndroid

**Comment by ditman** [comment]
@fangkyi03 this has been closed for >1 year, please [open a new issue](https://github.com/flutter/flutter/issues/new/choose) explaining in detail what you need :)

**Pull request merged**